Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift

```diff
@@ -0,0 +1,221 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftDiagnostics
+import SwiftParser
+import SwiftParserDiagnostics
+import SwiftSyntax
+import XCTest
+import _SwiftSyntaxTestSupport
+
+/// A typealias representing a location marker.
+///
+/// This string serves to pinpoint the exact location of a particular token in the Abstract Syntax Tree (AST).
+/// Once the token location is identified, it can be leveraged for various test-specific operations such as inserting diagnostics, notes, or fix-its,
+/// or for closer examination of the syntax tree.
+///
+/// Markers are instrumental in writing unit tests that require precise location data. They are commonly represented using emojis like 1️⃣, 2️⃣, 3️⃣, etc., to improve readability.
+///
+/// ### Example
+///
+/// In the following test code snippet, the emojis 1️⃣ and 2️⃣ are used as location markers:
+///
+/// ```swift
+/// func foo() -> Int {
+///   if 1️⃣1 != 0 2️⃣{
+///     return 0
+///   }
+///   return 1
+/// }
+/// ```
+typealias LocationMarker = String
```

Comment:
What do you think about the following? That way you can really specify which node you want to refer to at a specific position and you could also use `NodeLocator` to specify the highlighted nodes.

```swift
/// Refers to the first node in a syntax tree after `marker` that satisfies `match`.
///
/// For example, if you have
/// ```swift
/// var x: Int
/// 1️⃣func foo() {}
/// ```
///
/// Then `NodeLocator(marker: "1️⃣")` refers to the `CodeBlockItemSyntax` that
/// contains the function declaration and
/// `NodeLocator(marker: "1️⃣", matches: { $0.is(FunctionDeclSyntax.self) })`
/// refers to the function within.
struct NodeLocator {
  let marker: String
  let match: (Syntax) -> Bool = { _ in true }

  func findNode(in tree: Syntax) -> Syntax? {
    // TODO: Implement
  }
}
```

Reply:
I have a feeling this is going to be a bigger discussion. How about we move it out of the scope of this PR for now and discuss it in the issue: [Introduce Efficient Node Localization in Syntax Tree for Unit Tests](https://github.com/apple/swift-syntax/issues/2251)?

Reply:
Yes, sure!